src/MPGDCylinderBarrelTracker_geo.cpp

```diff
@@ -0,0 +1,563 @@
+           m_nam.c_str(), stave_width, det_length);
+
+  // ********** LAYER
+  // ***** ONE AND ONLY ONE LAYER
+  xml_coll_t li(x_det, _U(layer));
+  if (li.size() != 1) {
+    printout(ERROR, "MPGDCylinderBarrelTracker", "Number of layers = %d. Must be = 1",
+             (int)li.size());
+    throw runtime_error("Logics error in building modules.");
+  }
+  // ***** RETRIEVE PARAMETERS
+  xml_comp_t x_layer = li;
+  int lay_id         = x_layer.id();
+  if (x_layer.moduleStr() != m_nam) {
+    printout(ERROR, "MPGDCylinderBarrelTracker", "Layer \"%s\" does not match module \"%s\"",
+             x_layer.moduleStr().c_str(), m_nam.c_str());
+    throw runtime_error("Logics error in building layer.");
+  }
+  xml_comp_t x_barrel  = x_layer.child(_U(barrel_envelope));
+  double barrel_length = x_barrel.z_length();
+  double barrel_z0     = getAttrOrDefault(x_barrel, _U(z0), 0.);
+  // ***** LAYOUTS
+  xml_comp_t x_layout = x_layer.child(_U(rphi_layout));
+  double phi0         = x_layout.phi0();             // Starting phi of first module.
+  xml_comp_t z_layout = x_layer.child(_U(z_layout)); // Get the <z_layout> element.
+  double z_gap        = z_layout.gap();
+  // ***** UNVALID LAYOUT PROPERTIES
+  // # of staves (along phi) and sectors (along z) are now derived from stave
+  // width and sector length. Used to be specified directly. In order to remind
+  // the user this is no longer the case, let's forbid the use of the
+  // corresponding (and a few more) tags.
+  const int nUnvalids                     = 4;
+  const xml::Tag_t unvalidTags[nUnvalids] = {_U(phi_tilt), _U(nphi), _U(rc), _U(dr)};
+  for (int uv = 0; uv < nUnvalids; uv++) {
+    if (x_barrel.hasChild(unvalidTags[uv])) {
+      const string tag = _U(nphi);
+      printout(ERROR, "MPGDCylinderBarrelTracker",
+               "Layer \"%s\": Unvalid property \"%s\" in \"rphi_layout\"", m_nam.c_str(),
+               tag.c_str());
+      throw runtime_error("Logics error in building modules.");
+    }
+  }
+
+  // ***** MODELS
+  // Model = set of two radii of curvature used alternatingly for staves
+  //       + an offset to be applied radially.
+  // - The offset may be null if the two radii are different enough that
+  //  staves adjacent along phi do not overlap.
+  // - There needs to be 2, as of 2024/03: Inner and outer.
+  // StaveModel = radius of curvature (=> "nphi", given "rmin")
+  //            + offset
+  // sector2Models[2][2]: Which model for [inner,outer][r1,r2]
+  typedef struct {
+    string name;
+    double rmin;
+    // "rsensor" is supposed to have been set equal to the radius specified in
+    // the segmentation and is used here to double-check the consistency
+    // between the stack of module components and that segmentation radius.
+    double rsensor;
+    int nphi;
+    double offset;
+    int io;
+  } StaveModel;
+  StaveModel staveModels[4];
+  int sector2Models[2][2];
```

Comment:
Can this use a map?

Reply:
I would keep it as is for the time being, with fixed dimensions. The reason being that, although I tried to make it as general as possible, I haven't checked that more complicated settings, with more than two models or more than two distinct radii per model, are consistently dealt with later on in the code. Fixed dimensions act as a reminder of the fact.